# Retrieving Product Information using AWS Product Advertising API - Part 2
Now that the AWS Product Advertising API responses have been saved for each queried product(ASIN), the next step is to parse this information. The following fields will be needed for advertising purposes:
1. ASIN
2. Title
3. Brand
4. GTIN(s)
5. MPN/Model
6. Specifications/Details
7. Product Description
8. Features
9. Product Category
10. Stock Image(s)

Information from the AWS Product Advertising API responses will be used to fill as many fields as possible.

## Import Libraries, Packages, and Modules

In [1]:
# import libraries, packages, and modules
import pandas as pd # Data structures and data analysis tools library
from bs4 import BeautifulSoup # HTML and XML document parsing package
from re import search, sub

## Construct pandas DataFrame from Tab-Seperated Values File of AWS Product Advertising API Responses by ASIN

In [2]:
# construct pandas DataFrame with data from tsv file of aws pa api responses by asin
column_number_with_ASIN = 0 # set this variable to the numberical value of the csv column containing the desired ASINs
asin_responses_df = pd.read_csv('aws_pa_item_attributes_responses.tsv', header=0, sep='\t')
print(asin_responses_df.shape) # print DataFrame Shape
asin_responses_df.head() # preview DataFrame

(107, 5)


asin                                item_attributes_xml  \
0  B004XC6GJ0  <?xml version="1.0" ?><html><body><itemlookupr...   
1  B000P1DEHU  <?xml version="1.0" ?><html><body><itemlookupr...   
2  B004NEUJKA  <?xml version="1.0" ?><html><body><itemlookupr...   
3  B0011ULQNI  <?xml version="1.0" ?><html><body><itemlookupr...   
4  B001O0DP48  <?xml version="1.0" ?><html><body><itemlookupr...   

                                editorial_review_xml  \
0  <?xml version="1.0" ?><html><body><itemlookupr...   
1  <?xml version="1.0" ?><html><body><itemlookupr...   
2  <?xml version="1.0" ?><html><body><itemlookupr...   
3  <?xml version="1.0" ?><html><body><itemlookupr...   
4  <?xml version="1.0" ?><html><body><itemlookupr...   

                                          images_xml        date  
0  <?xml version="1.0" ?><html><body><itemlookupr...  2019-01-29  
1  <?xml version="1.0" ?><html><body><itemlookupr...  2019-01-29  
2  <?xml version="1.0" ?><html><body><itemlookupr...  2019-01-29  
3  <?xml version="1.0" ?><html><body><itemlookupr...  2019-01-29  
4  <?xml version="1.0" ?><html><body><itemlookupr...  2019-01-29

## Construct pandas DataFrame for the Desired Product Fields

In [3]:
aws_pa_product_data_df = pd.DataFrame(columns=['asin', 'title', 'brand', 'gtin', 'mpn_model', 'specs', 'description', 'features', 'category', 'image'])
aws_pa_product_data_df.head() # return dataframe head

Empty DataFrame
Columns: [asin, title, brand, gtin, mpn_model, specs, description, features, category, image]
Index: []

## Define Functions for Parsing XML Data

In [4]:
# define function for parsing xml data to prevent script from breaking on error, returns blank string instead
def parse_xml(xml, elementstring):
    """
    Parses product attributes from xml data.
    """
    xml = BeautifulSoup(xml, 'lxml')
    try:
        if elementstring == 'image':
            return(xml.item.largeimage.url.text)
        elif elementstring == 'description':
            try:
                desc_string = xml.find_all('content')[0].get_text()
                return(desc_string[0:search('<', str(image_test.find_all('content')[0].text)).span()[0]])
            except:
                return(xml.find_all('content')[0].get_text())
        elif elementstring == 'features':
            feature_list = []
            for feature in xml.find_all('feature'):
                if 'warranty' in feature.text.lower():
                    pass
                elif 'report to amazon' in feature.text.lower():
                    pass
                else:
                    feature_list.append(feature.text)
            if len(feature_list) > 0:
                return(feature_list)
            else:
                return('')
        elif elementstring == 'mpn':
            model_mpn1 = str(xml.find_all('model')[0].text)
            model_mpn2 = str(xml.find_all('mpn')[0].text)
            if model_mpn1 != '':
                return(model_mpn1)
            else:
                return(model_mpn2)
        elif 'list' in elementstring:
            list_elements = []
            for list_item in xml.find(elementstring).findAll(elementstring + 'element', recursive=False):
                list_elements.append(list_item.text)
            return(list_elements)
        elif len(xml.find_all(elementstring)) > 1:
            list_elements = []
            for element in xml.find_all(elementstring):
                list_elements.append(element.text)
            return(list_elements)
        else:
            return(str(xml.find_all(elementstring)[0].text))
    except:
        return('')

In [5]:
# define function for parsing specifications from xml data to prevent script from breaking on error, returns blank string instead
def spec_parse_xml(xml):
    """
    Parses product specifications from xml data.
    """
    try:
        xml = BeautifulSoup(xml, 'lxml')
        field_spec_list = []
        specs_map_dict = {}
        for element in xml.itemattributes.children:
            if 'sku' in element.name.lower() or 'warranty' in element.name.lower() or 'feature' in element.name.lower() or 'legaldisclaimer' in element.name.lower() or 'language' in element.name.lower() or 'department' in element.name.lower() or 'dimensions' in element.name.lower() or 'title' in element.name.lower() or 'product' in element.name.lower():
                    pass
            elif 'listprice' in element.name.lower():
                pass
            elif 'list' in element.name.lower():
                specs_map_dict[element.name] = 'list'
            elif len(xml.find_all(element.name)) > 1:
                specs_map_dict[element.name] = 'multi'
            else:
                specs_map_dict[element.name] = 'single'
        for attrib in specs_map_dict:
            specs_value_list = []
            if specs_map_dict[attrib] == 'list':
                if attrib == 'catalognumberlist':
                    for cn in xml.catalognumberlist.children:
                        specs_value_list.append(cn.text)
                    specs_map_dict[attrib] = list(set(specs_value_list))
                elif attrib == 'eanlist':
                    for cn in xml.eanlist.children:
                        specs_value_list.append(cn.text)
                    specs_map_dict[attrib] = list(set(specs_value_list))
                elif attrib == 'upclist':
                    for cn in xml.eanlist.children:
                        specs_value_list.append(cn.text)
                    specs_map_dict[attrib] = list(set(specs_value_list))
            elif specs_map_dict[attrib] == 'multi':
                for value in xml.find_all(attrib):
                    specs_value_list.append(value.text)
                specs_map_dict[attrib] = list(set(specs_value_list))
            elif specs_map_dict[attrib] == 'single':
                specs_map_dict[attrib] = xml.find_all(attrib)[0].text
            else:
                pass
        for key in specs_map_dict:
            dict_string = ''
            dict_string += key + ': '
            if type(specs_map_dict[key]) == list:
                    for i, list_items in enumerate(specs_map_dict[key]):
                        if i < len(specs_map_dict[key]) - 1:
                            dict_string += list_items + ', '
                        else:
                            dict_string += list_items
            else:
                dict_string += specs_map_dict[key]
            field_spec_list.append(dict_string)
        return(field_spec_list)
    except:
        return('')

In [6]:
# define function for construction product category taxonomy
def cat_taxonomy(productgroup, producttype):
    """
    Constructs product taxonomy from provided arguments
    """
    cat_string = '{} > {}'.format(productgroup, producttype)
    if cat_string == ' > ':
        return('')
    else:
        return(cat_string)

In [7]:
# construct product class
class product:
    """
    Class constructs product object and defines the objects product attributes
    """
    def __init__(self, item_attributes_xml, editorial_review_xml, images_xml):
        self.ia_xml = item_attributes_xml
        self.i_xml = images_xml
        self.er_xml = editorial_review_xml
        self.image = parse_xml(self.i_xml, 'image')
        self.description = parse_xml(self.er_xml, 'description')
        self.asin = parse_xml(self.ia_xml, 'asin')
        self.title = parse_xml(self.ia_xml, 'title')
        self.brand = parse_xml(self.ia_xml, 'brand')
        self.upc = parse_xml(self.ia_xml, 'upc')
        self.ean = parse_xml(self.ia_xml, 'ean')
        self.isbn = parse_xml(self.ia_xml, 'isbn')
        self.mpn = parse_xml(self.ia_xml, 'mpn')
        self.model = parse_xml(self.ia_xml, 'model')
        self.features = parse_xml(self.ia_xml, 'features')
        self.department = parse_xml(self.ia_xml, 'department')
        self.group = parse_xml(self.ia_xml, 'productgroup')
        self.type = parse_xml(self.ia_xml, 'producttypename')
        self.specs = spec_parse_xml(self.ia_xml)

In [8]:
# function for iterating through each product
def parse_loop(df, column_names):
    """
    1. Takes DataFrame of product xml data.
    2. Iterates through rows creating an object of the product class.
    3. Constructs a dictionary from the object attributes.
    4. Returns a DataFrame of desired product fields.
    """
    new_df = pd.DataFrame(columns=column_names)
    for row in df.iterrows():
        catalog_dict = {}
        catalog_item = product(row[1][1], row[1][2], row[1][3])
        catalog_dict = {
            'asin' : row[1][0],
            'title' : catalog_item.title,
            'brand' : catalog_item.brand,
            'gtin' : catalog_item.ean,
            'mpn_model' : catalog_item.mpn,
            'specs' : catalog_item.specs,
            'description' : catalog_item.description,
            'features' : catalog_item.features,
            'category' : cat_taxonomy(catalog_item.group, catalog_item.type),
            'image' : catalog_item.image,
        }
        new_df = new_df.append(catalog_dict, ignore_index=True)
    return(new_df)

## Construct DataFrame of Product Data

In [9]:
# Construct DataFrame of Product Data 
aws_pa_product_data_df = parse_loop(asin_responses_df, aws_pa_product_data_df.columns)
aws_pa_product_data_df

asin                                              title  \
0    B004XC6GJ0  ARRIS SURFboard SB6121 4x4 DOCSIS 3.0 Cable Mo...   
1    B000P1DEHU  Sony DVD/VCR Progressive Scan Combo Player SLV...   
2    B004NEUJKA  Uniden D1660-2 DECT6.0 Caller ID Cordless hand...   
3    B0011ULQNI                                                      
4    B001O0DP48  Uniden DECT 6.0 Silver Cordless Phone with Cal...   
5    B00DQV2BDO          Panasonic KX-TGA402N 1.9GHz Extra Handset   
6    B000E1B2SO                                                      
7    B00005K3B1  Uniden EXA2950 900 MHz Cordless Phone with Dig...   
8    B00000J0E6    RCA VHS Rewinder (Discontinued by Manufacturer)   
9    B0001B86GI  Uniden DCX640 2.4 GHz Accessory Handset for DC...   
10   B075X4G51V               Kinyo AW600A VHS Video Tape Rewinder   
11   B006N9ZZS4  Polycom VVX 500 12-line Business Media Phone P...   
12   B003A1LK1O  eMachines PS/2 Multimedia Black & Silver Keybo...   
13   B00NIAYOX8  RCA Remote R130c1 for L32wd12 L32wd12yx7 L37wd...   
14   B00IO3AQC2  Orion "OEM" 076N0EA010 VCR2511 VCR2510C VHSA66...   
15   B0028MD7F8   LG Electronics/Zenith AKB69680401 REMOTE CONTROL   
16   B001DFH0C2           Sony Remote Commander (RM-Y906), RM-Y906   
17   B001DFPR9A  SONY 1-480-097-11 REMOTE CONTROL RM-AAU013 OEM...   
18   B00A4U1XVQ   Panasonic PVQV201 PVV402 PVV4021 LSSQ0263 Remote   
19   B0091HQHZU  Genuine HP 0957-2259 AC Power Supply Adapter +...   
20   B00WLJLYJY  Power Supply AC ADAPTER 22V 455mA 3Pin for HP ...   
21   B000IJEZI6  Philips SWS2102W/17 RF Modulator Video Convert...   
22   B00PVV543G  Texas Instruments TI-36X Pro Scientific Calcul...   
23   B077ZKF9XK  Lenovo New Genuine ThinkPad Yoga 45 Watt AC Ad...   
24   B0036RH2U0           NETGEAR XAV2001 Powerline AV 200 Adapter   
25   B004FWOECY             TomTom XL 335SE 4.3-Inch GPS Navigator   
26   B0057OCS9A                                                      
27   B0015F0A8G                                                      
28   B000046S41                           Nintendo 64 Transfer Pak   
29   B004RZJ60O                    Polaroid Impulse Instant Camera   
..          ...                                                ...   
77   6300247163                            The Turning Point [VHS]   
78   B000056VP3                          Remember the Titans [VHS]   
79   6302415853                           Dances With Wolves [VHS]   
80   B000EFNVB6              SONY T120V - Sony S-VHS Videocassette   
81   6301914090                              Days of Thunder [VHS]   
82   6302428122                          Bill Cosby, Himself [VHS]   
83   0783219105                                Happy Gilmore [VHS]   
84   0790739305                    Why Do Fools Fall in Love [VHS]   
85   0790738074                                The Music Man [VHS]   
86   6302226899                      Best Years of Our Lives [VHS]   
87   B000068CNG  Panasonic KX-TG2700S 2.4 GHz DSS Expandable Co...   
88   B00008Y2EN  Uniden EXP3241 2.4 GHz Extended Range Cordless...   
89   B001DFOBIS                Sony 1-418-694-11 REMOTE(RMT-V266A)   
90   B001QPT1PI                                Homedics Ac Adapter   
91   B003BNY7R0  VTech 6305 Accessory Cordless Handset, Silver/...   
92   B00KZPD0D4      Genuine Hon-Kwang D0660 AC Adapter 6VDC 600mA   
93   B005GY3GWO  Fuji AC-5VW AC Power Adapter for Fujifilm Digi...   
94   B00997KEZQ  Pitney Bowes Power Supply AC Adapter 24VDC 1.5...   
95   B0001BRIBM   Belkin ERGOBOARD PRO KEYBOARD BLACK (F8E887-BLK)   
96   B0000950SC  Panasonic KX-TG5110M 5.8 GHz DSS Expandable Co...   
97   B001ELLMGU        Sony RM-SC3 Factory Original Remote Control   
98   B001K5NSYE           Sony 148057011 REMOTE CONTROL (RMADU007)   
99   B0000C7AFN                       Minolta Vectis 40 APS Camera   
100  B002VR6A9K  Microsoft Arc Wireless Keyboard for PC and Xbo...   
101  B000F49ZHU  Panasonic KX-TG5623B Black 5.8 GHz FHSS GigaRa...   
102  B0073W703I  Pana

## Save aws_pa_product_data_df state in a Tab-delimited values file
As this process may be repeated over time and existing data may remain useful, the mode is set to 'a' for append.

In [10]:
# save as tsv
aws_pa_product_data_df.to_csv('aws_pa_product_data.tsv', sep='\t' , index=False, mode='a')
print('Export complete.')

Export complete.


## Construct List of Products Missing the Title Field

In [11]:
# Construct List of Products Missing the Title Field
aws_pa_product_data_missing_df = aws_pa_product_data_df['asin'][aws_pa_product_data_df['title'] == '']
aws_pa_product_data_missing_df

3     B0011ULQNI
6     B000E1B2SO
26    B0057OCS9A
27    B0015F0A8G
42    B00006IIOL
44    B0007Z70YM
46    B000Q8UPKQ
Name: asin, dtype: object

## Results

In [12]:
# Number of Products with Title Present
total = aws_pa_product_data_df.shape[0]
title_present = aws_pa_product_data_df['asin'][aws_pa_product_data_df['title'] != ''].shape[0]
title_absent = aws_pa_product_data_df['asin'][aws_pa_product_data_df['title'] == ''].shape[0]
print('Number of Product Rows with Title Present: ' + str(title_present))
print('Number of Product Rows with Title Absent: ' + str(title_absent))
print('Total Number of Product Rows: ' + str(total))
print('Success Rate: ' + str(round(title_present / total * 100, 2)) + '%')

Number of Product Rows with Title Present: 100
Number of Product Rows with Title Absent: 7
Total Number of Product Rows: 107
Success Rate: 93.46%


## Discussion

This was the second part of a two part project. The goal of parsing the desired product fields was largely successful. As depicted in the <i>aws_pa_product_data_missing_df</i> DataFrame, data about some products is unavailable for retrieval via the Amazon Web Services Product Advertising API. Consequently, data for these products will need to be obtained from a different source. One alternative source that may be useful is eBay. A future project may address these shortcomings in the interest of improving data quality and completeness. This project was successful with at  93.46% of products now being matched with at least a product title.